### DISCUSSION TITLE
Clarity on code executed outside of steps

### DISCUSSION THREAD

**Discussion by michaelhays** [inngest/inngest#586] [Q&A] [answered]
Hello, I'm reading the documentation on [Multi-Step Functions](https://www.inngest.com/docs/functions/multi-step), and wanted to make sure I understand how the code run outside of steps is invoked.

Take the following Inngest function:

```ts
export type InngestUpdateUserEvent = {
  data: { userId: string }
}

const inngestUpdateUser = inngest.createFunction(
  { name: 'User > Update' },
  { event: 'user.update' },
  async ({ event, step }) => {
    const user = await getUser(event.data.userId)

    await step.run('Update user name', async () => {
      await saveUser({ ...user, name: 'New Name' })
    })

    await step.run('Email user', async () => {
      await sendEmail(user, `Your name has been updated to ${user.name}`)
    })
  },
)
```

Is `const user = await getUser(event.data.userId)` run once (at the start of the event), or twice (at the start of each step)?

**Comment by tonyhb** [comment, marked as answer]
It runs many times.  Steps run once and the first run's output is stored in function state and injected on the next run.

Code outside of steps is ran multiple times _up until_ we execute the next step.  For example, if this await was at the end of the function it only runs once, as it's not executed until all steps finish.  However, being at the top it's ran every time a step is invoked.

**Reply by michaelhays** [reply]
Got it, thanks! Your updated docs (especially the [Gotchas](https://www.inngest.com/docs/functions/multi-step#gotchas)) is an improvement in explaining the functionality there.

**Reply by ShravanSunder** [reply]
@tonyhb is nested steps possible and are they memoized? 

Would the below cause any issues or does the framework work with the below?

```
export default inngest.createFunction(
  { id: "activation-email" },
  { event: "app/user.created" },
  async ({ event, step }) => {
    await step.run("send-welcome-email", async () => {
       await step.run("do something", async () => { do something });
       await step.run("do something else", async () => { do something else });
      return await sendEmail({ email: event.user.email, template: "welcome" });
    });

...
```
